subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/VersionSelectionReasons.java

```diff
@@ -149,7 +150,21 @@ public boolean equals(Object o) {
                 return false;
             }
             DefaultComponentSelectionReason that = (DefaultComponentSelectionReason) o;
-            return Objects.equal(descriptions, that.descriptions);
+            return sameDescriptorsAs(that);
+        }
+
+        private boolean sameDescriptorsAs(DefaultComponentSelectionReason that) {
```

Comment:
Why not use an arraylist?

Reply:
Because of the use of `getLast`.

Reply:
Not sure I get the problem. `get(list.size -1)`.

Reply:
It's not a problem, it's just more efficient. `ArrayList` and `ArrayDeque` are almost the same. Except that here I don't need random access, but need access to the last element, which `ArrayDeque` knows directly.

Reply:
I doubt that changes anything here, it's two field accesses instead of one. But you could get rid of the custom equals method.